retrofit2/src/main/java/com/linecorp/armeria/client/http/retrofit2/ArmeriaRetrofit.java

```diff
@@ -18,72 +18,104 @@
-
-import com.google.common.annotations.VisibleForTesting;
-import com.google.common.escape.Escaper;
-import com.google.common.escape.Escapers;
+import java.util.regex.Pattern;
 
 import com.linecorp.armeria.client.http.HttpClient;
 import com.linecorp.armeria.common.Scheme;
 import com.linecorp.armeria.common.SessionProtocol;
 
 import okhttp3.HttpUrl;
 import retrofit2.Retrofit;
+import retrofit2.Retrofit.Builder;
 
 /**
  * A helper class for creating a new {@link Retrofit} instance with {@link ArmeriaCallFactory}.
  * For example,
  * <pre>{@code
- * HttpClient httpClient = Clients.newClient(ClientFactory.DEFAULT,
- *                                           "none+http://localhost:8080",
- *                                           HttpClient.class);
  *
- * Retrofit retrofit = ArmeriaRetrofit.builder(httpClient)
+ * Retrofit retrofit = ArmeriaRetrofit.builder("none+http://localhost:8080")
  *     .addConverterFactory(GsonConverterFactory.create())
  *     .build();
  *
  * MyApi api = retrofit.create(MyApi.class);
  * Response&lt;User&gt; user = api.getUser().execute();
  * }
  * </pre>
+ *
+ * ArmeriaRetrofit even supports EndpointGroup, so you can create Retrofit like below,
+ * <pre>{@code
+ *
+ * EndpointGroupRegistry.register("foo",
+ *                                new StaticEndpointGroup(Endpoint.of("127.0.0.1", 8080)),
+ *                                ROUND_ROBIN);
+ *
+ * Retrofit retrofit = ArmeriaRetrofit.builder("none+http://group:foo")
+ *     .addConverterFactory(GsonConverterFactory.create())
+ *     .build();
+ * }
+ * </pre>
+ * ArmeriaRetrofit will convert http://group:foo to http://__group__foo internally to avoid okHttp3's
+ * {@link HttpUrl} throwing exception when parsing http://group:foo . To avoid your client have possibility to
+ * access domain which starts with __group__, you can use {@link ArmeriaRetrofit#builder(String, String)} to
+ * customize the internal behavior.
  */
 public final class ArmeriaRetrofit {
-    private static final Escaper GROUP_NAME_ESCAPER = Escapers.builder()
-                                                              .addEscape('@', "_")
-                                                              .addEscape('/', "_")
-                                                              .addEscape('\\', "_")
-                                                              .addEscape('?', "_")
-                                                              .addEscape('#', "_")
-                                                              .addEscape(':', "_")
-                                                              .build();
+
+    private static final Pattern GROUP_PREFIX_PATTERN = Pattern.compile("^[_0-9a-z]+$");
+    private static final String DEFAULT_GROUP_PREFIX = "__group__";
 
     /**
      * Creates a {@link Retrofit.Builder} with {@link ArmeriaCallFactory} using the specified
      * {@link HttpClient} instance.
+     *
+     * @deprecated Use {@link String} instead.
```

Comment:
Having a convenience method taking String is nice, but there are probably good use cases for passing an HttpClient itself instead, like setting deadline or decorators for metrics, right?